src/Toml/Bi/Code.hs

```diff
@@ -0,0 +1,72 @@
+module Toml.Bi.Code
+       ( -- * Types
+         BiToml
+       , Env
+       , St
+
+         -- * Exceptions
+       , DecodeException (..)
+       , prettyException
+
+         -- * Encode/Decode
+       , decode
+       , encode
+       ) where
+
+import Control.Monad.Except (ExceptT, runExceptT)
+import Control.Monad.Reader (Reader, runReader)
+import Control.Monad.State (State, execState)
+import Data.Bifunctor (first)
+import Data.Semigroup ((<>))
+import Data.Text (Text)
+import Data.Foldable (toList)
+
+import Toml.Bi.Monad (Bi, Bijection (..))
+import Toml.Parser (ParseException (..), parse)
+import Toml.PrefixTree (Key (..), unPiece)
+import Toml.Printer (prettyToml)
+import Toml.Type (TOML (..), ValueType, showType)
+
+import qualified Data.Text as Text
+
+-- | Type of exception for converting from 'Toml' to user custom data type.
+data DecodeException
+    = KeyNotFound Key  -- ^ No such key
+    | TableNotFound Key  -- ^ No such table
+    | TypeMismatch Key Text ValueType -- ^ Expected type vs actual type
+    | ParseError ParseException  -- ^ Exception during parsing
+    deriving (Eq, Show)  -- TODO: manual pretty show instances
```

Comment:
Isn't `show` is just `toString . prettyException` now?

Reply:
Personally I like to keep default definition of `Show` just to be able to see structure of data type if needed. But I don't mind writing `Show` using `prettyException`. Though, it requires more code than just generating `Show` instance...